drivers/flash/flash_ifx_cat1.c

```diff
@@ -0,0 +1,190 @@
+/*
+ * Copyright 2023 Cypress Semiconductor Corporation (an Infineon company) or
+ * an affiliate of Cypress Semiconductor Corporation
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#define DT_DRV_COMPAT	  infineon_cat1_flash_controller
+#define SOC_NV_FLASH_NODE DT_PARENT(DT_INST(0, fixed_partitions))
+
+#define PAGE_LEN DT_PROP(SOC_NV_FLASH_NODE, erase_block_size)
+
+#include <zephyr/kernel.h>
+#include <zephyr/devicetree.h>
+#include <zephyr/drivers/flash.h>
+#include <zephyr/logging/log.h>
+
+#include "cyhal_flash.h"
+
+LOG_MODULE_REGISTER(flash_infineon_cat1, CONFIG_FLASH_LOG_LEVEL);
+
+/* Device config structure */
+struct ifx_cat1_flash_config {
+	uint32_t base_addr;
+	uint32_t max_addr;
+};
+
+/* Data structure */
+struct ifx_cat1_flash_data {
+	cyhal_flash_t flash_obj;
+	struct k_sem sem;
+};
+
+static struct flash_parameters ifx_cat1_flash_parameters = {
+	.write_block_size = DT_PROP(SOC_NV_FLASH_NODE, write_block_size),
+	.erase_value = 0x00,
+};
+
+static inline void flash_ifx_sem_take(const struct device *dev)
+{
+	struct ifx_cat1_flash_data *data = dev->data;
+
+	k_sem_take(&data->sem, K_FOREVER);
+}
+
+static inline void flash_ifx_sem_give(const struct device *dev)
+{
+	struct ifx_cat1_flash_data *data = dev->data;
+
+	k_sem_give(&data->sem);
+}
+
+static int ifx_cat1_flash_read(const struct device *dev, off_t address, void *buffer, size_t length)
+{
+	struct ifx_cat1_flash_data *data = dev->data;
+	const struct ifx_cat1_flash_config *config = dev->config;
+
+	flash_ifx_sem_take(dev);
+
+	cy_rslt_t rslt = cyhal_flash_read(&data->flash_obj, address + config->base_addr,
+					  (uint8_t *)buffer, length);
+	if (rslt != CY_RSLT_SUCCESS) {
+		flash_ifx_sem_give(dev);
+		return -EIO;
+	}
```

Comment:
Worth at least `LOG_DBG` or `LOG_ERR` the `rslt` here, unless the `cyhal_flash_read` does some logging.
If you have checked everything to this point and the only reason for this function to fail is actual problem somewhere under `cyhal_flash_read`, returning just -EIO gives no info for developer.